js/ui/searchbar.js

```diff
@@ -0,0 +1,178 @@
+    },
+
+    // todo add regex support
+    // todo add search history
+    // we don't have to worry about the files blacklist because they are already removed from the project structure
+    search: function() {
+      var self = this;
+      var filesToSearch = 0; // we track this so we know when we're done
+      var matches = 0;
+      var maxMatches = this.maxMatches;
+
+      var isCaseSensitive = this.element.find("#search-case-check").checked;
+      var displayQuery = this.input.value;
+      var searchQuery = isCaseSensitive ? displayQuery : displayQuery.toUpperCase();
+      var resultsTab = sessions.addFile("Search Results for:\n" + displayQuery + "\n");
+      resultsTab.fileName = "Results: " + displayQuery;
+
+      // searchDirectory() and searchFile() are defined inside search() because they affect the local matches variable
+      function searchDirectory(node) {
+        for (var i = 0; i < node.children.length; i++) {
+          var childNode = node.children[i];
+          if (childNode.isDirectory) {
+            searchDirectory(childNode);
+          } else if (childNode.entry.isFile) {
+            filesToSearch++;
+            searchFile(childNode);
+          }
+        }
+      }
+
+      function searchFile(node) {
+        var prevLine = "";
+
+        chrome.fileSystem.getDisplayPath(node.entry, function(path) {
+          var file = new File(node.entry);
+            if (matches >= maxMatches) {
+              return;
+            }
+
+            file.read(function(err, data) {
+              var lines = data.split("\n");
+              var line, msg;
+              var firstFind = true;
+              var printedLines = {}; // only print each line once per file per search
+
+              for (var i = 0; i < lines.length && matches < maxMatches; i++) {
+                compareLine = isCaseSensitive ? lines[i] : lines[i].toUpperCase();
+                if (compareLine.indexOf(searchQuery) > -1) {
+                  matches++;
+                  msg = "";
+                  if (!printedLines[i] && !printedLines[i-1]) { // only add break if it and the line before it have not been printed
+                    if (firstFind) {
+                      msg += "\n" + node.entry.fullPath + "\n";
+                      firstFind = false;
+                    } else {
+                      msg += "...\n";
+                    }
+                    msg += self.formatResultCode(i-1, prevLine);
+                    msg += self.formatResultCode(i, lines[i]);
+                    printedLines[i] = true;
+                  }
+
+                  if (i < lines.length - 1) {
+                    msg += self.formatResultCode(i+1, lines[i+1]);
+                    printedLines[i+1] = true;
+                  }
+                  self.appendToResults(resultsTab, msg);
+                }
+                prevLine = lines[i];
+              }
+
+              if (--filesToSearch === 0) { // we have finished searching all files
```

Comment:
Actually, it looks like it does always run, it just takes a really long time after it stops printing results if the project is very large. This is related to the short circuiting not being done quite right. I'll fix it tonight.